File: src/models/balance.py
"""잔고 데이터 모델."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class StockHolding(BaseModel):
    ticker: str                # pdno
    name: str                  # prdt_name
    quantity: int              # hldg_qty
    sellable_quantity: int     # ord_psbl_qty
    avg_price: float           # pchs_avg_pric
    purchase_amount: int       # pchs_amt
    current_price: int         # prpr
    eval_amount: int           # evlu_amt
    eval_profit_loss: int      # evlu_pfls_amt
    eval_profit_rate: float    # evlu_pfls_rt
    # J1 (2026-05-11) — stock_master(CTPF1002R 캐시) 조인.
    # 미캐시/조회 예외 시 None — 프론트는 "확인중" 라벨 노출.
    nxt_tradable: Optional[bool] = None
    krx_halted: Optional[bool] = None
    excg_dvsn_cd: Optional[str] = None


class AccountSummary(BaseModel):
    deposit: int               # dnca_tot_amt (예수금총금액)
    stock_eval_amount: int     # scts_evlu_amt (유가평가금액)
    total_eval_amount: int     # tot_evlu_amt (총평가금액)
    net_asset: int             # nass_amt (순자산금액)
    purchase_total: int        # pchs_amt_smtl_amt (매입금액합계)
    eval_total: int            # evlu_amt_smtl_amt (평가금액합계)
    profit_loss_total: int     # evlu_pfls_smtl_amt (평가손익합계)


class BuyableInfo(BaseModel):
    cash_available: int        # ord_psbl_cash
    max_buy_amount: int        # max_buy_amt
    max_buy_quantity: int      # max_buy_qty

File: frontend/src/types/balance.ts
export interface Holding {
  ticker: string
  name: string
  quantity: number
  sellable_quantity: number
  avg_price: number
  purchase_amount: number
  current_price: number
  eval_amount: number
  eval_profit_loss: number
  eval_profit_rate: number
  // J1 (2026-05-11) — stock_master 캐시 조인 결과.
  // 미캐시/조회 예외 시 null/undefined — UI 는 "확인중" 표시.
  nxt_tradable?: boolean | null
  krx_halted?: boolean | null
  excg_dvsn_cd?: string | null
  // 2026-08-04 — 업종 한글명(bstp_kor_isnm) 우선 → KRX 산업지수 플래그/업종코드
  // 폴백 → `미분류-{ticker}`. 백엔드 `sector_naming` 단일 진실원.
  sector?: string | null
  // cycle339 — 그 종목을 보유한 전략의 청산선. 백엔드 `position_exit_lines` 단일 진실원.
  // 🔴 판정 불가는 전부 null 이고 화면은 `—` 를 그린다 — 숫자를 지어내지 않는다.
  strategy_id?: string | null
  stop_price?: number | null
  /** `effective` = 전략 실효 손절선(check_exit 와 동일 산식) / `hard_pct` = 고정%손절 근사. */
  stop_source?: 'effective' | 'hard_pct' | null
  target_price?: number | null
  /** `measured_move` = BFB 측정 목표(부분 익절 트리거). 그 밖 전략은 목표가가 없다. */
  target_source?: 'measured_move' | null
}

export interface BalanceSummary {
  deposit: number
  stock_eval_amount: number
  total_eval_amount: number
  net_asset: number
  purchase_total: number
  eval_total: number
  profit_loss_total: number
}

export interface BalanceData {
  holdings: Holding[]
  summary: BalanceSummary
}
